tardis/io/logger/logger.py

```diff
@@ -1,165 +1,265 @@
+                logger.removeHandler(handler)
+        
+        self.logger.addHandler(widget_handler)
+        logging.getLogger("py.warnings").addHandler(widget_handler)
+        
 
-    # Creating a list for Storing all the Loggers which are derived from TARDIS
-    tardis_loggers = tardis_logger()
+def get_environment():
+    """Determine the execution environment"""
+    try:
+        import IPython
+        ipython = IPython.get_ipython()
+        
+        if ipython is None:
+            return 'standard'
+            
+        # Check for VSCode specific environment variables
+        if any(x for x in ('VSCODE_PID', 'VSCODE') if x in os.environ):
+            return 'vscode'
+            
+        # Check if running in Jupyter notebook
+        if 'IPKernelApp' in ipython.config:
+            return 'jupyter'
+            
+        return 'standard'
+    except:
+        return 'standard'
 
-    if logging_level in LOGGING_LEVELS:
-        for logger in tardis_loggers:
-            logger.setLevel(LOGGING_LEVELS[logging_level])
+class AsyncEmitLogHandler(logging.Handler):
+    def __init__(self, log_outputs, colors):
+        super().__init__()
+        self.log_outputs = log_outputs
+        self.colors = colors
+        self.environment = get_environment()
+        self.main_thread_id = threading.get_ident()
+        self.futures = []
+        
+        # Single event loop and thread for all environments
+        self.loop = asyncio.new_event_loop()
+        self.thread = threading.Thread(target=self._run_event_loop, daemon=True)
+        self.thread.start()
+        
+        # Only set up display handle for Jupyter
+        if self.environment == 'jupyter':
+            self.display_handle = display(logger_widget, display_id=True)
 
-    if logger.filters:
-        for filter in logger.filters:
-            for logger in tardis_loggers:
-                logger.removeFilter(filter)
-
-    if specific_log_level:
-        filter_log = FilterLog(LOGGING_LEVELS[logging_level])
-        for logger in tardis_loggers:
-            logger.addFilter(filter_log)
-    else:
-        for filter in logger.filters:
-            for logger in tardis_loggers:
-                logger.removeFilter(filter)
-
-
-def tardis_logger():
-    """
-    Generates the list of the loggers which are derived from TARDIS
-    All loggers which are of the form `tardis.module_name` are added to the list
-
-    Parameters
-    ----------
-    list_for_loggers : list
-        List for storing the loggers derived from TARDIS
-
-    Returns
-    -------
-    list_for_loggers : list
-    """
-    list_for_loggers = []
-    for name in logging.root.manager.loggerDict:
-        if not name.find("tardis"):
-            list_for_loggers.append(logging.getLogger(name))
-    return list_for_loggers
+    def _run_event_loop(self):
+        """Runs event loop in separate thread"""
+        asyncio.set_event_loop(self.loop)
+        self.loop.run_forever()
```

Comment:
Anything that is forever is slightly concerning. Does this thread persist after the simulation completes?

Reply:
There is a close() method where this is cleaned up properly. I added a few more things to ensure it was properly cleaned up in https://github.com/tardis-sn/tardis/pull/2872/commits/3e9c38d27c1f1e5192dad8c29cb7fbdaf28b2236

https://docs.python.org/3/library/logging.html#logging.Handler.close